### PR TITLE
iox-#2041 Move test class for ctor and assignment operator to hoofs testing

### LINKED ISSUE #2041: Move test class for ctor and assignment operators to hoofs testing
## Brief feature description

The test class from the `iox::vector` was duplicated for the `iox::list` and `iox::forward_list`. In order to remove the code duplication and make it available for other tests the class should be refactored out into a separate file in hoofs testing.

## Tasks
- [x] Make test class to separate file
- [x] Find a better name

### PR DESCRIPTION
**Pull request by elBoberido**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] All touched (C/C++) source code files from `iceoryx_hoofs` are added to `./clang-tidy-diff-scans.txt`
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Notes for Reviewer

I'm currently trying to replace the keep-alive mechanism from an IPC message to a timestamp in the shared memory. For this I needed to refactor the `FixedPositionContainer` and this in turn needs a test class for cTor and assignment operators. This PR moves the `CTorTest` class from the `vector` tests into a location where it can be reused by other tests.

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] All touched (C/C++) source code files from `iceoryx_hoofs` have been added to `./clang-tidy-diff-scans.txt`
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Relates to #2041

### CODE REVIEW

**Comment by elBoberido** [comment]
@FerdinandSpitzschnueffler @mossmaurice @dkroenke @MatthiasKillat @elfenpiff feel free to add yourself as reviewer

**Review by FerdinandSpitzschnueffler — APPROVED** [review]

**Comment by elBoberido** [comment]
@FerdinandSpitzschnueffler muchas gracias :)

File: iceoryx_hoofs/testing/include/iceoryx_hoofs/testing/ctor_and_assignment_operator_test_class.hpp [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+#include <cstdint>
+#include <vector>
+
+namespace iox
+{
+namespace testing
+{
+template <typename T = uint64_t, T DEFAULT_VALUE = 0>
+class CTorAndAssignmentOperatorTestClass
```
Comment (mossmaurice) [inline iceoryx_hoofs/testing/include/iceoryx_hoofs/testing/ctor_and_assignment_operator_test_class.hpp:30]: @elBoberido I think one of those names would fit better:

1. `LifetimeVerificator`
2. `LifetimeTracker`
3. `CTorAssignmentCounterClass`
4. `StatisticalObjectMonitor`
5. `ObjectLifetimeInspector`

What do you think?

**Review by mossmaurice — APPROVED** [review]
I'm approving this PR as it already has one approval. However, I'm no a big fan of the name of the new class `CTorAndAssignmentOperatorTestClass`. Please check my proposals. Can also be adapted in a follow-up PR.

File: iceoryx_hoofs/testing/include/iceoryx_hoofs/testing/ctor_and_assignment_operator_test_class.hpp (hunk 1 above)
Comment (elBoberido) [reply iceoryx_hoofs/testing/include/iceoryx_hoofs/testing/ctor_and_assignment_operator_test_class.hpp:30]: I'm also not that happy with the name. I just labeled it according what it does :sweat_smile: 

What do you think of `LifetimeAndAssignmentTracker`?

I'll merge now but will happily change in a follow up

Comment (elBoberido) [reply iceoryx_hoofs/testing/include/iceoryx_hoofs/testing/ctor_and_assignment_operator_test_class.hpp:30]: Move discussion to the issue :)

Comment (mossmaurice) [reply iceoryx_hoofs/testing/include/iceoryx_hoofs/testing/ctor_and_assignment_operator_test_class.hpp:30]: @elBoberido 

> What do you think of `LifetimeAndAssignmentTracker`?

Sounds good to me. Way better than what is currently used :stuck_out_tongue:

**Pull request merged**